bodhi/server/models.py

```diff
@@ -2213,42 +2213,61 @@ def get_test_gating_info(self):
+            satisfied_requirements.extend(response.get('satisfied_requirements', []))
+            unsatisfied_requirements.extend(response.get('unsatisfied_requirements', []))
+
+        return (satisfied_requirements, unsatisfied_requirements)
+
     def _get_test_gating_status(self):
         """
         Query Greenwave about this update and return the information retrieved.
 
         Returns:
             TestGatingStatus:
                 - TestGatingStatus.ignored if no tests are required
-                - TestGatingStatus.failed if policies are not satisfied
-                - TestGatingStatus.passed if policies are satisfied, and there
-                  are required tests
+                - TestGatingStatus.passed if there are required tests and policies are satisfied
+                - TestGatingStatus.waiting if there are required tests that have not yet completed,
+                  no required test has failed, and update was last modified less than 2 hours ago
+                - TestGatingStatus.failed otherwise (failed required tests, missing required
+                  test results and last modified more than 2 hours ago)
 
         Raises:
             BodhiException: When the ``greenwave_api_url`` is undefined in configuration.
-            RuntimeError: If Greenwave did not give us a 200 code.
+            RuntimeError: If Greenwave did not give us a 200 code, including if no applicable
+            policies were found.
         """
-        # If an unrestricted policy is applied and no tests are required
-        # on this update, let's set the test gating as ignored in Bodhi.
-        status = TestGatingStatus.ignored
-        for data in self.greenwave_request_batches(verbose=False):
-            response = util.greenwave_api_post(self._greenwave_api_url, data)
-            if not response['policies_satisfied']:
-                return TestGatingStatus.failed
+        (satisfied, unsatisfied) = self._greenwave_requirements
 
-            if status != TestGatingStatus.ignored or response['summary'] != 'no tests are required':
-                status = TestGatingStatus.passed
+        if not satisfied and not unsatisfied:
+            return TestGatingStatus.ignored
 
-        return status
+        if satisfied and not unsatisfied:
+            return TestGatingStatus.passed
 
-    @property
-    def _unsatisfied_requirements(self):
-        unsatisfied_requirements = []
-        for data in self.greenwave_request_batches(verbose=False):
-            response = util.greenwave_api_post(self._greenwave_api_url, data)
-            unsatisfied_requirements.extend(response['unsatisfied_requirements'])
+        if datetime.utcnow() - self.last_modified < timedelta(hours=2):
+            if all(req.get('type', '') == 'test-result-missing' for req in unsatisfied):
+                return TestGatingStatus.waiting
 
-        return unsatisfied_requirements
+        return TestGatingStatus.failed
```

Comment:
Not sure if it is worth optimizing for this case, but it could stop fetching any next "greenwave_request_batches" when there is an unsatisfied requirement (that is not `test-result-missing` when last modification is under 2hrs).

I'm assuming that it won't happen too often so it is fine as is.

Reply:
Hmm, I see what you mean. The trade-off there is that we'd probably have to put back the near-duplication I took out in this version, unless I can think of a clever way to do it. I can think about it a bit.